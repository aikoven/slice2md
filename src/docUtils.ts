import {escapeRegExp} from 'lodash';

/** @internal */
export type DocComponent =
  | TextDocComponent
  | TagDocComponent
  | InlineTagDocComponent;

/** @internal */
export type TextDocComponent = {
  type: 'text';
  value: string;
};

/** @internal */
export type TagDocComponent = {
  type: 'tag';
  tag: string;
  value?: string;
};

/** @internal */
export type InlineTagDocComponent = {
  type: 'inline-tag';
  tag: string;
  value: string;
};

/** @internal */
export function parseDoc(doc: string): DocComponent[] {
  const [text, ...tags] = doc.split(/(?=^@\w+)/gm);

  const components: DocComponent[] = [];

  const inlineTagRegExp = /\s*{@(\w+)\s+([^}]+)}\s*/g;

  while (true) {
    const {lastIndex} = inlineTagRegExp;
    const match = inlineTagRegExp.exec(text);

    if (match != null) {
      components.push(
        {
          type: 'text',
          value: text.slice(lastIndex, match.index).trim(),
        },
        {
          type: 'inline-tag',
          tag: match[1],
          value: match[2],
        },
      );
    } else {
      components.push({
        type: 'text',
        value: text.slice(lastIndex, text.length).trim(),
      });
      break;
    }
  }

  for (const tag of tags) {
    const match = tag.match(/^@(\w+)(?:\s+(.*))?$/s);

    if (match) {
      const [, tag, value] = match;
      components.push({
        type: 'tag',
        tag,
        value: value == null ? value : value.trim(),
      });
    }
  }

  return components;
}

/** @internal */
export function findTag(
  tags: TagDocComponent[],
  name: string,
): string | undefined {
  for (const {value} of tags) {
    if (value != null) {
      const match = value.match(
        new RegExp(`^${escapeRegExp(name)}\\s+(.*)$`, 's'),
      );

      if (match) {
        return match[1].trim();
      }
    }
  }

  return undefined;
}
