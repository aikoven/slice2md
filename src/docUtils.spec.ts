import dedent = require('dedent');
import {parseDoc} from './docUtils';

test('parseDoc', () => {
  expect(
    parseDoc(
      dedent(`
        Text before inline tag {@first inline content} more text {@second lol}
        another line

        @tag1 tag with single line
        @tag2 tag with
        multiple
        lines
        @tag3
      `),
    ),
  ).toMatchSnapshot();
});
