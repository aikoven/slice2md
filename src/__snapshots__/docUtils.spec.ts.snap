// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseDoc 1`] = `
Array [
  Object {
    "type": "text",
    "value": "Text before inline tag",
  },
  Object {
    "tag": "first",
    "type": "inline-tag",
    "value": "inline content",
  },
  Object {
    "type": "text",
    "value": "more text",
  },
  Object {
    "tag": "second",
    "type": "inline-tag",
    "value": "lol",
  },
  Object {
    "type": "text",
    "value": "another line",
  },
  Object {
    "tag": "tag1",
    "type": "tag",
    "value": "tag with single line",
  },
  Object {
    "tag": "tag2",
    "type": "tag",
    "value": "tag with
multiple
lines",
  },
  Object {
    "tag": "tag3",
    "type": "tag",
    "value": undefined,
  },
]
`;
